### ISSUE TITLE
'lxc stop' starts hanging after some time of LXD daemon's hard work

### ISSUE THREAD

**Issue by Akint** [lxc/lxd#4217]
* Distribution: Ubuntu
 * Distribution version: 16.04 (xenial)
```
# lxc info
config:
  core.https_address: '[::]:8443'
  images.compression_algorithm: pbzip2
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - x.x.x.x:8443
  - y.y.y.y:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    ...
    -----END CERTIFICATE-----
  certificate_fingerprint: 9d138df151b84d54a9d3583dab6859759910911922481c3afc40b07291dae84f
  driver: lxc
  driver_version: 2.0.8
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-104-generic
  server: lxd
  server_pid: 13090
  server_version: "2.21"
  storage: zfs | dir
  storage_version: 0.6.5.6-0ubuntu16 | 1
```

I have couple of VMs acting as Jenkins nodes for building LXD images. Containers are constantly being created, configured, published and deleted, mostly several containers in parallel.
After some undetermined time of such work a node stops serving LXD commands. `systemctl restart lxd` usually solves the problem, but sometimes it hangs as well and I have to reboot the node.
Today it hung again and I had a couple of minutes to investigate. This time it acts like this: you can create container, but when you try to stop it with `lxc stop <container-name>`, it hangs. You can take a look at `lxc list` (terminate command execution with Ctrl+C or open another session) to see the container has actually been stopped.
`/var/log/lxd/<container-name>/lxc.log` says:
```
lxc_commands - commands.c:lxc_cmd_rsp_recv:177 - Command get_cgroup failed to receive response: Connection reset by peer.
```
`strace` says `lxc stop` process waits on futex:
```
strace -fp $(pgrep -f 'lxc stop')
strace: Process 26142 attached with 6 threads
[pid 26142] futex(0xe482d0, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 26147] futex(0xc420178110, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 26146] futex(0xe673e0, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 26145] futex(0xc420034b90, FUTEX_WAIT, 0, NULL <unfinished ...>
[pid 26144] epoll_wait(4,  <unfinished ...>
[pid 26143] restart_syscall(<... resuming interrupted futex ...>
```
Output of `lxc stop --debug <container-name>` and `lxc monitor` are attached to the issue.
[lxd_monitor.log](https://github.com/lxc/lxd/files/1689075/lxd_monitor.log)
[lxd_stop.log](https://github.com/lxc/lxd/files/1689076/lxd_stop.log)

**Comment by stgraber** [comment]
Are you seeing any lxc-monitord processes on your system? If so, does killing that process help?
Also, does passing --force to `lxc stop` fix this issue?

When calling `lxc stop` without --force, all that LXD does is send a signal to the container's init process, then wait for it to turn itself off. That's very similar to sending an ACPI event to your VM for a clean shutdown.

So it can be that for some reason your init system in the container is misbehaving too. Trying with --force would clarify that.

**Comment by Akint** [comment]
Yes, I'm seeing `lxc-monitord` process. Killing that process makes `lxc stop` execution fail with an error "shutting down the container failed", but after that start/stop operations go smoothly. `lxc-monitord` is running again after that.
Since killing `lxc-monitord` worked, I haven't had a chance to try `lxc stop --force`.
Is `lxc-monitord`'s misbehavior some kind of known fact and are there any workarounds to prevent it from hanging?

**Comment by stgraber** [comment]
Yeah, lxc-monitord is known to occasionally hang, in which case killing it usually unsticks whatever's going on. We've fixed that particular issue in liblxc 2.0.9, unfortunately most distros aren't yet on that version and so still suffer from this issue.

In your case, an option would be to switch from the LXD deb that you're using (out of backports it looks like) and switch to the LXD snap instead. The snap bundles a much more recent version of liblxc which actually entirely did away with lxc-monitord.

You may want to try that on one of your systems, the migration should just be:

 - snap install lxd
 - lxd.migrate

Which will walk you through moving all existing data from the deb to the snap and subsequently removing the deb from your system.

Closing this issue since it's a liblxc issue which has already been corrected upstream.
